stdlib/asm/math/u64.masm

```diff
@@ -673,6 +676,52 @@ export.unchecked_shr
     cswap
 end
 
+# Performs left shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where (d,c) = a << b, 
+# which d contains the bits shifted out.
+# This takes 20 cycles.
+export.overflowing_shl
+    pow2
+    u32split
+    exec.overflowing_mul
+end
+
+# Performs right shift of one unsigned 64-bit integer preserving the overflow and
+# using the pow2 operation.
+# The input value to be shifted is assumed to be represented using 32 bit limbs.
+# The shift value is assumed to be in the range [0, 64).
+# Stack transition looks as follows:
+# [b, a_hi, a_lo, ...] -> [d_hi, d_lo, c_hi, c_lo, ...], where c = a >> b, d = a << (64 - b).
+# This takes 66 cycles.
+export.overflowing_shr
+    push.64             # (64 - b)
+    dup.1               
+    sub
+    movdn.3             # move result keeping [b, a_hi, a_lo, ....]
+    
+    dup.2               # dup a_lo, a_hi and bring b
+    dup.2
+    dup.2
+    exec.unchecked_shr  # c = a >> b
+
+    movup.4             # bring duplicated a_lo and a_hi
+    movup.4
+    movup.5             # bring the (64 - b) and b
+    movup.5
+    padw                # adding zeroes on 0, 1, 2 and 3 to be able to use cdropw
+    dup.4               # dup b
+    eq.0
+    cdropw              # if b is 0, swap the four zeroes added with b, (64 - b), a_hi, a_lo
+                        # after that, drop 0, 1, 2 and 3
+    drop                # drop the last added 0 or dup b to keep the format [b, a_hi, a_lo, ....]
+
+    exec.unchecked_shl  # d = a << (64 - b)
```

Comment:
What do you think about this approach? If it works, I think it takes up 63 cycles (though, would be good to double-check):
```
push.64
dup.1
sub
dup.3
dup.3
dup.3
exec.unchecked_shr
movdn.5
movdn.5
padw
pad
movup.6
eq.0
cdrowp
drop
exec.unchecked_shl
```